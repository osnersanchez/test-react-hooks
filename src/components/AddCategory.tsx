import React, { ChangeEvent, Dispatch, FormEvent, SetStateAction, useState } from 'react';

interface PropsTypes {
  setCategories: Dispatch<SetStateAction<string[]>>
}

const AddCategory = ({ setCategories }: PropsTypes) => {
  const [inputValue, setInputValue] = useState<string>('');

  const handleinputChange = (e: ChangeEvent<HTMLInputElement>) => setInputValue(e.target.value);
  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories((categories: string[]) => [inputValue, ...categories]);
      setInputValue('');
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        value={inputValue}
        onChange={handleinputChange}
        type="text"
      />
    </form>
  )
};

export default AddCategory;