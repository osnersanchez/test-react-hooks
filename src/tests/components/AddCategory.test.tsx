import { HTMLAttributes, shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import AddCategory from '../../components/AddCategory';

const expectValue = 'The Pepe';
const evetInputMock = { target: { value: expectValue } };
const evetFormMock = { preventDefault() { } };

describe('AddCategory test suit', () => {
  let wrapper: ShallowWrapper
  let setCategoryMock: jest.Mock;

  beforeEach(() => {
    setCategoryMock = jest.fn<void, string[]>()
    wrapper = shallow(<AddCategory setCategories={setCategoryMock} />);
  })

  test('should render default template when default properties are set', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should changed value input when setInputValue method is called', () => {
    const inputRef: ShallowWrapper<HTMLAttributes> = wrapper.find('input');
    inputRef.simulate('change', evetInputMock);

    expect(wrapper.find('p').text().trim()).toBe(expectValue);
  });

  test('should not call setCategoryMock method when submit method is called and input length is less than 2', () => {
    const formRef: ShallowWrapper<HTMLAttributes> = wrapper.find('form');
    formRef.simulate('submit', evetFormMock);

    expect(setCategoryMock).not.toHaveBeenCalled();
  });

  test('should call setCategoryMock method when submit method is called and input length is greater than 2', () => {
    const inputRef: ShallowWrapper<HTMLAttributes> = wrapper.find('input');
    inputRef.simulate('change', evetInputMock);
    wrapper.find('form').simulate('submit', evetFormMock);

    expect(setCategoryMock).toHaveBeenCalledWith(expect.any(Function));
    expect(inputRef.prop('value')).toBe('');
  });
})

