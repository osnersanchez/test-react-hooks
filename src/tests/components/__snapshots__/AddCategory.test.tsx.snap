// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory test suit should render default template when default properties are set 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
